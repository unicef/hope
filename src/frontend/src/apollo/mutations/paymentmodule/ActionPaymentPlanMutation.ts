import { gql } from '@apollo/client';

export const ACTION_PAYMENT_PLAN = gql`
  mutation ActionPP($input: ActionPaymentPlanInput!) {
    actionPaymentPlanMutation(input: $input) {
      paymentPlan {
        id
        status
      }
    }
  }
`;
