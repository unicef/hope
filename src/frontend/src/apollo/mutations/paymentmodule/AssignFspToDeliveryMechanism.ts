import { gql } from '@apollo/client';

export const ASSIGN_FSP_TO_DELIVERY_MECHANISM = gql`
  mutation AssignFspToDeliveryMech($input: AssignFspToDeliveryMechanismInput!) {
    assignFspToDeliveryMechanism(input: $input) {
      paymentPlan {
        id
        deliveryMechanisms {
          id
          name
          fsp {
            id
            name
          }
        }
      }
    }
  }
`;
