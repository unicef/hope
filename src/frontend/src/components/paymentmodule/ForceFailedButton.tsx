import { Box, Button, DialogContent, DialogTitle } from '@mui/material';
import * as React from 'react';
import { useState } from 'react';
import { useTranslation } from 'react-i18next';
import { Dialog } from '@containers/dialogs/Dialog';
import { DialogActions } from '@containers/dialogs/DialogActions';
import { DialogContainer } from '@containers/dialogs/DialogContainer';
import { DialogFooter } from '@containers/dialogs/DialogFooter';
import { DialogTitleWrapper } from '@containers/dialogs/DialogTitleWrapper';
import { useSnackbar } from '@hooks/useSnackBar';
import { useMarkPayAsFailedMutation } from '@generated/graphql';

export interface ForceFailedButtonProps {
  paymentId: string;
  disabled?: boolean;
}
export function ForceFailedButton({
  paymentId,
  disabled = false,
}: ForceFailedButtonProps): React.ReactElement {
  const { t } = useTranslation();
  const [isOpenModal, setOpenModal] = useState(false);
  const { showMessage } = useSnackbar();
  const [mutate, { loading }] = useMarkPayAsFailedMutation();

  const submit = async (): Promise<void> => {
    try {
      await mutate({
        variables: {
          paymentId,
        },
      });
      setOpenModal(false);
      showMessage(t('Payment has been marked as failed.'));
    } catch (e) {
      e.graphQLErrors.map((x) => showMessage(x.message));
    }
  };

  return (
    <Box>
      <Box p={2}>
        <Button
          color="primary"
          variant="contained"
          onClick={() => setOpenModal(true)}
          data-cy="button-mark-as-failed"
          disabled={disabled}
        >
          {t('Mark as failed')}
        </Button>
      </Box>
      <Dialog
        open={isOpenModal}
        onClose={() => setOpenModal(false)}
        scroll="paper"
        aria-labelledby="form-dialog-title"
        maxWidth="md"
      >
        <DialogTitleWrapper>
          <DialogTitle>{t('Mark as failed')}</DialogTitle>
        </DialogTitleWrapper>
        <DialogContent>
          <DialogContainer>
            <Box p={5}>
              {t('Are you sure you would like to mark payment as failed?')}
            </Box>
          </DialogContainer>
        </DialogContent>
        <DialogFooter>
          <DialogActions>
            <Button onClick={() => setOpenModal(false)}>{t('CANCEL')}</Button>
            <Button
              type="submit"
              color="primary"
              variant="contained"
              onClick={() => submit()}
              data-cy="button-submit"
              disabled={loading}
            >
              {t('Mark as failed')}
            </Button>
          </DialogActions>
        </DialogFooter>
      </Dialog>
    </Box>
  );
}
