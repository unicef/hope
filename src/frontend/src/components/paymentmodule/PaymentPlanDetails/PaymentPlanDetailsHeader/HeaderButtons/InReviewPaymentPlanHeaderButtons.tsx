import { Box } from '@mui/material';
import * as React from 'react';
import { PaymentPlanQuery } from '@generated/graphql';
import { MarkAsReleasedPaymentPlan } from '../MarkAsReleasedPaymentPlan';
import { RejectPaymentPlan } from '../RejectPaymentPlan';

export interface InReviewPaymentPlanHeaderButtonsProps {
  paymentPlan: PaymentPlanQuery['paymentPlan'];
  canReject: boolean;
  canMarkAsReleased: boolean;
}

export function InReviewPaymentPlanHeaderButtons({
  paymentPlan,
  canReject,
  canMarkAsReleased,
}: InReviewPaymentPlanHeaderButtonsProps): React.ReactElement {
  return (
    <Box display="flex" alignItems="center">
      {canReject && <RejectPaymentPlan paymentPlanId={paymentPlan.id} />}
      {canMarkAsReleased && (
        <MarkAsReleasedPaymentPlan paymentPlan={paymentPlan} />
      )}
    </Box>
  );
}
