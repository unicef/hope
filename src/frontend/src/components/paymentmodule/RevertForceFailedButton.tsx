import { Box, Button, DialogContent, DialogTitle } from '@mui/material';
import * as React from 'react';
import { useState } from 'react';
import { useTranslation } from 'react-i18next';
import { Field, Form, Formik } from 'formik';
import * as Yup from 'yup';
import CalendarTodayRoundedIcon from '@mui/icons-material/CalendarTodayRounded';
import { Dialog } from '@containers/dialogs/Dialog';
import { DialogActions } from '@containers/dialogs/DialogActions';
import { DialogFooter } from '@containers/dialogs/DialogFooter';
import { DialogTitleWrapper } from '@containers/dialogs/DialogTitleWrapper';
import { useSnackbar } from '@hooks/useSnackBar';
import { FormikTextField } from '@shared/Formik/FormikTextField';
import { FormikDateField } from '@shared/Formik/FormikDateField';
import { useRevertMarkPayAsFailedMutation } from '@generated/graphql';

export interface RevertForceFailedButtonProps {
  paymentId: string;
  disabled?: boolean;
}
export function RevertForceFailedButton({
  paymentId,
  disabled = false,
}: RevertForceFailedButtonProps): React.ReactElement {
  const { t } = useTranslation();
  const [isOpenModal, setOpenModal] = useState(false);
  const { showMessage } = useSnackbar();
  const [mutate, { loading }] = useRevertMarkPayAsFailedMutation();

  const validationSchema = Yup.object().shape({
    deliveredQuantity: Yup.number()
      .min(0)
      .max(99999999, t('Number is too big')),
    deliveryDate: Yup.date().required(t('Delivery date is required')),
  });

  const submit = async (values, { resetForm }): Promise<void> => {
    try {
      await mutate({
        variables: {
          paymentId,
          deliveredQuantity: values.deliveredQuantity,
          deliveryDate: values.deliveryDate,
        },
      });
      setOpenModal(false);
      showMessage(t('Force failed has been reverted.'));
      resetForm();
    } catch (e) {
      e.graphQLErrors.map((x) => showMessage(x.message));
    }
  };

  return (
    <Box>
      <Box p={2}>
        <Button
          color="primary"
          variant="contained"
          onClick={() => setOpenModal(true)}
          data-cy="button-revert-mark-as-failed"
          disabled={disabled}
        >
          {t('Revert mark as failed')}
        </Button>
      </Box>
      <Formik
        initialValues={{
          deliveredQuantity: 0,
          deliveryDate: '',
        }}
        validationSchema={validationSchema}
        onSubmit={submit}
      >
        {({ submitForm, resetForm }) => (
          <Dialog
            open={isOpenModal}
            onClose={() => setOpenModal(false)}
            scroll="paper"
            aria-labelledby="form-dialog-title"
            maxWidth="md"
          >
            <DialogTitleWrapper>
              <DialogTitle>{t('Revert mark as failed')}</DialogTitle>
            </DialogTitleWrapper>
            <DialogContent>
              <Form>
                <Field
                  name="deliveredQuantity"
                  label={t('Delivered Quantity')}
                  type="number"
                  required
                  fullWidth
                  variant="outlined"
                  component={FormikTextField}
                />
                <Field
                  name="deliveryDate"
                  label={t('Delivery Date')}
                  component={FormikDateField}
                  required
                  fullWidth
                  decoratorEnd={<CalendarTodayRoundedIcon color="disabled" />}
                />
              </Form>
            </DialogContent>
            <DialogFooter>
              <DialogActions>
                <Button
                  onClick={() => {
                    setOpenModal(false);
                    resetForm();
                  }}
                >
                  {t('CANCEL')}
                </Button>
                <Button
                  type="submit"
                  color="primary"
                  variant="contained"
                  onClick={submitForm}
                  data-cy="button-submit"
                  disabled={loading}
                >
                  {t('Revert mark as failed')}
                </Button>
              </DialogActions>
            </DialogFooter>
          </Dialog>
        )}
      </Formik>
    </Box>
  );
}
