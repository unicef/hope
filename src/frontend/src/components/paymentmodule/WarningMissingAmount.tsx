import { Box } from '@mui/material';
import WarningIcon from '@mui/icons-material/Warning';
import { useTranslation } from 'react-i18next';
import styled from 'styled-components';
import { ReactElement } from 'react';

const WarnIcon = styled(WarningIcon)`
  color: ${({ theme }) => theme.hctPalette.orange};
`;

interface WarningMissingAmountProps {
  amount: number;
  currency: string;
}

export function WarningMissingAmount({
  amount,
  currency,
}: WarningMissingAmountProps): ReactElement {
  const { t } = useTranslation();

  return (
    <Box display="flex" alignItems="center">
      <WarnIcon />
      {t('Missing')} {amount} {currency}
    </Box>
  );
}
