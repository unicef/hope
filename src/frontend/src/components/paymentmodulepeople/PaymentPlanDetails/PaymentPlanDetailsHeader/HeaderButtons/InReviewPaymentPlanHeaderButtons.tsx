import { Box } from '@mui/material';
import { MarkAsReleasedPaymentPlan } from '../MarkAsReleasedPaymentPlan';
import { RejectPaymentPlan } from '../RejectPaymentPlan';
import { ReactElement } from 'react';
import { PaymentPlanDetail } from '@restgenerated/models/PaymentPlanDetail';

export interface InReviewPaymentPlanHeaderButtonsProps {
  paymentPlan: PaymentPlanDetail;
  canReject: boolean;
  canMarkAsReleased: boolean;
}

export function InReviewPaymentPlanHeaderButtons({
  paymentPlan,
  canReject,
  canMarkAsReleased,
}: InReviewPaymentPlanHeaderButtonsProps): ReactElement {
  return (
    <Box display="flex" alignItems="center">
      {canReject && <RejectPaymentPlan paymentPlanId={paymentPlan.id} />}
      {canMarkAsReleased && (
        <MarkAsReleasedPaymentPlan paymentPlan={paymentPlan} />
      )}
    </Box>
  );
}
