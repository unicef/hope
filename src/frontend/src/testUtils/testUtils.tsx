import { ReactElement } from 'react';
import { BrowserRouter } from 'react-router-dom';
import { render, RenderOptions } from '@testing-library/react';
import { Formik } from 'formik';
import noop from 'lodash/noop';
import { TestProviders } from './testProviders';
import { ProgramStatus } from '@generated/graphql';
import { MockLink } from '@apollo/client/testing';

const customRender = (
  ui: ReactElement,
  options?: Omit<RenderOptions, 'queries'>,
) =>
  render(
    <>
      <BrowserRouter>
        <Formik initialValues={{}} onSubmit={noop}>
          {ui}
        </Formik>
      </BrowserRouter>
    </>,
    {
      wrapper: TestProviders,
      ...options,
    },
  );

export * from '@testing-library/react';
export { customRender as render };

let m_w = 123456789;
let m_z = 987654321;
const mask = 0xffffffff;

// Takes any integer
export const seed = (i) => {
  m_w = (123456789 + i) & mask;
  m_z = (987654321 - i) & mask;
};

// Returns number between 0 (inclusive) and 1.0 (exclusive),
// just like Math.random().
export const random = () => {
  m_z = (36969 * (m_z & 65535) + (m_z >> 16)) & mask;
  m_w = (18000 * (m_w & 65535) + (m_w >> 16)) & mask;
  let result = ((m_z << 16) + (m_w & 65535)) >>> 0;
  result /= 4294967296;
  return result;
};

export class ApolloLoadingLink extends MockLink {
  constructor() {
    super([]);
  }

  request() {
    return null;
  }
}

export const fakeContextProgram = {
  selectedProgram: {
    id: '1',
    name: 'someName',
    status: ProgramStatus.Active,
    pduFields: null,
    dataCollectingType: {
      id: '1',
      householdFiltersAvailable: true,
      individualFiltersAvailable: true,
      label: 'data collecting type',
      code: '123',
      type: 'full',
      children: null,
    },
  },
  setSelectedProgram: () => {},
  isActiveProgram: true,
  isSocialDctType: false,
  isStandardDctType: true,
  programHasPdu: false,
};
