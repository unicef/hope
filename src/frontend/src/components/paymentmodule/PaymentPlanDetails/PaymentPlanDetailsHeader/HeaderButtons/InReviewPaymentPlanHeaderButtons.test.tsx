import { MockedProvider } from '@apollo/react-testing';
import * as React from 'react';
import { act } from 'react-dom/test-utils';
import wait from 'waait';
import { fakeApolloPaymentPlan } from '../../../../../../fixtures/paymentmodule/fakeApolloPaymentPlan';
import { fakeActionPpMutation } from '../../../../../../fixtures/paymentmodule/fakeApolloActionPaymentPlanMutation';
import { render } from '../../../../../testUtils/testUtils';
import { InReviewPaymentPlanHeaderButtons } from './InReviewPaymentPlanHeaderButtons';

describe('components/paymentmodule/PaymentPlanDetails/PaymentPlanDetailsHeader/HeaderButtons/InReviewPaymentPlanHeaderButtons', () => {
  it('should render with buttons', async () => {
    const { container } = render(
      <MockedProvider addTypename={false} mocks={fakeActionPpMutation}>
        <InReviewPaymentPlanHeaderButtons
          paymentPlan={fakeApolloPaymentPlan}
          canReject
          canMarkAsReleased
        />
      </MockedProvider>,
    );

    await act(() => wait(0)); // wait for the mutation to complete

    expect(container).toMatchSnapshot();
  });
});
