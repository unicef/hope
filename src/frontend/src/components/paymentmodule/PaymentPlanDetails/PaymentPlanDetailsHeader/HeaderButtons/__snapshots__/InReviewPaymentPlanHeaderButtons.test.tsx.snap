// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components/paymentmodule/PaymentPlanDetails/PaymentPlanDetailsHeader/HeaderButtons/InReviewPaymentPlanHeaderButtons should render with buttons 1`] = `
<div>
  <div
    class="MuiBox-root css-70qvj9"
  >
    <div
      class="MuiBox-root css-hpgf8j"
    >
      <button
        class="MuiButtonBase-root MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium MuiButton-colorPrimary MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium MuiButton-colorPrimary css-1cl7z2v-MuiButtonBase-root-MuiButton-root"
        data-cy="button-reject"
        tabindex="0"
        type="button"
      >
        Reject
      </button>
    </div>
    <div
      class="MuiBox-root css-hpgf8j"
    >
      <button
        class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-colorPrimary MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-colorPrimary css-cx92lf-MuiButtonBase-root-MuiButton-root"
        data-cy="button-mark-as-released"
        tabindex="0"
        type="button"
      >
        Mark as released
      </button>
    </div>
  </div>
</div>
`;
